build-tools/src/main/java/software/amazon/awssdk/buildtools/findbugs/ToBuilderIsCorrect.java

```diff
@@ -0,0 +1,459 @@
+     * constructor to copy over fields.
+     *
+     * Key 1: The buildable FQCN, Key 2: the invoked builder FQCN, Values: the signature of the constructor invoked by the
+     * toBuilder method.
+     */
+    private final Map<String, Map<String, String>> constructorsInvokedFromToBuilder = new HashMap<>();
+
+    /**
+     * Which builder constructors reference which other builder constructors. This is used when a buildable uses a builder's
+     * constructor to copy over fields, and that constructor delegates some copying to another internal constructor.
+     *
+     * Key 1: The builder FQCN, Key 2: the invoking constructor signature, Value: the target constructor signature.
+     */
+    private final Map<String, Map<String, String>> crossConstructorReferences = new HashMap<>();
+
+    /**
+     * Which classes or interfaces a builder extends. This is used to determine what builder is being modified when it is
+     * invoked virtually from a toBuilder implementation.
+     *
+     * Key: The builder FQCN, Value: the FQCN parent classes and interfaces (including transitive)
+     */
+    private final Map<String, Set<String>> builderParents = new HashMap<>();
+
+    /**
+     * Whether the current class being assessed is a builder.
+     */
+    private boolean isBuilder = false;
+
+    /**
+     * Whether the current class being assessed is a buildable.
+     */
+    private boolean isBuildable = false;
+
+    /**
+     * Whether the current method being assessed is a builder's constructor.
+     */
+    private boolean inBuilderConstructor = false;
+
+    /**
+     * Whether the current method being assessed is a buildable's toBuilder.
+     */
+    private boolean inBuildableToBuilder = false;
+
+    public ToBuilderIsCorrect(BugReporter bugReporter) {
+        this.bugReporter = bugReporter;
+    }
+
+    /**
+     * Determine whether we should visit the provided class. We only visit builders and buildables.
+     */
+    @Override
+    public boolean shouldVisit(JavaClass obj) {
+        isBuilder = false;
+        isBuildable = false;
+        inBuilderConstructor = false;
+        inBuildableToBuilder = false;
+
+        try {
+            if (obj.isInterface()) {
+                return false;
+            }
+
+            boolean result = false;
+            JavaClass[] parentInterfaces = obj.getAllInterfaces();
+            for (JavaClass i : parentInterfaces) {
+                if (i.getClassName().equals("software.amazon.awssdk.utils.builder.CopyableBuilder")) {
+                    builders.put(obj.getClassName(), obj);
+                    builderParents.put(obj.getClassName(),
+                                       Stream.concat(Arrays.stream(parentInterfaces), Arrays.stream(obj.getSuperClasses()))
+                                             .map(JavaClass::getClassName)
+                                             .collect(Collectors.toSet()));
```

Comment:
It looks like the same code will get executed for any matching `i`. Is there a way to break it out of the loop?

Reply:
Nice catch, yes!

Reply:
Technically it might be slower this way with it outside of the loop, since most classes won't have a matching value and even fewer will have multiple matching values.